### PR TITLE
feat: added-tests + split parser and downloader +  update logger

### PR DESCRIPTION
**Pull request by Qas1modo**

### CODE REVIEW

File: FundParser/BL/Services/DownloaderService/IDownloaderService.cs [hunk 1]
```diff
@@ -0,0 +1,8 @@
+﻿namespace FundParser.BL.Services.DownloaderService
+{
+    public interface IDownloaderService
+    {
+        public Task<string?> DownloadTextFileAsStringAsync(string url,
```
Comment (adlermichal) [inline FundParser/BL/Services/DownloaderService/IDownloaderService.cs]: Remove redundant visibility modifier

File: FundParser/BL/DTOs/AddCompanyDTO.cs [hunk 2]
```diff
@@ -2,9 +2,9 @@ namespace FundParser.BL.DTOs;
 
 public class AddCompanyDTO
 {
-    public string Name { get; set; }
+    public string Name { get; set; } = null!;
```
Comment (adlermichal) [inline FundParser/BL/DTOs/AddCompanyDTO.cs]: I am guessing this was done to suppress nullable warnings, but I don't think this is the best way to do it.
Personally, I would add a required modifier instead and it will be also safer, since it won't actually be null when we don't expect it..

File: FundParser/BL/DTOs/AddCompanyDTO.cs [hunk 3]
```diff
@@ -2,9 +2,9 @@ namespace FundParser.BL.DTOs;
...
 
-    public string Cusip { get; set; }
+    public string Cusip { get; set; } = null!;
 
-    public string Ticker { get; set; }
+    public string Ticker { get; set; } = null!;
```
Comment (adlermichal) [inline FundParser/BL/DTOs/AddCompanyDTO.cs]: I think the ticker can actually be null. Or maybe it's just an empty string? Not sure, but one company has that field always empty.

File: FundParser/BL/Services/CsvParserService/CsvParserService.cs [hunk 4]
```diff
@@ -0,0 +1,40 @@
+using System.Globalization;
+using CsvHelper;
+using CsvHelper.Configuration;
+using FundParser.BL.Services.LoggingService;
+
+namespace FundParser.BL.Services.CsvParserService;
+
+public class CsvParserService<T> : ICsvParserService<T>
```
Comment (adlermichal) [inline FundParser/BL/Services/CsvParserService/CsvParserService.cs]: nit: I would go with CsvParsingService or just CsvParser

File: FundParser/BL/Services/CsvParserService/CsvParserService.cs [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+    public IEnumerable<T>? ParseString(string inputString, CancellationToken cancellationToken = default)
+    {
+        try
+        {
+            using var reader = new StringReader(inputString);
+            var columnCount = typeof(T).GetProperties().Length;
+            _csvConfig.ShouldSkipRecord = r => r.Row.ColumnCount != columnCount;
+            using var csvReader = new CsvReader(reader, _csvConfig);
+            return csvReader.GetRecords<T>().ToList();
+        } catch (Exception e)
```
Comment (adlermichal) [inline FundParser/BL/Services/CsvParserService/CsvParserService.cs]: nit: formatting (catch on new line), but I'll run reformat before submitting, so I guess we don't have to deal with it now.

File: FundParser/BL/Services/DownloaderService/DownloaderService.cs [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+            _client = new HttpClient(messageHandler);
+            _logger = logger;
+        }
+
+
+        public async Task<string?> DownloadTextFileAsStringAsync(string url, IEnumerable<(string, string)>? headers = default, CancellationToken cancellationToken = default)
+        {
+            try
+            {
+                HttpRequestMessage request = new(HttpMethod.Get, new Uri(url));
```
Comment (adlermichal) [inline FundParser/BL/Services/DownloaderService/DownloaderService.cs]: I believe we are using `var`s through most of the codebase, so we should stick to that

File: FundParser/BL/Services/CsvParserService/CsvParserService.cs [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+
+    public IEnumerable<T>? ParseString(string inputString, CancellationToken cancellationToken = default)
+    {
+        try
+        {
+            using var reader = new StringReader(inputString);
+            var columnCount = typeof(T).GetProperties().Length;
+            _csvConfig.ShouldSkipRecord = r => r.Row.ColumnCount != columnCount;
+            using var csvReader = new CsvReader(reader, _csvConfig);
+            return csvReader.GetRecords<T>().ToList();
```
Comment (adlermichal) [inline FundParser/BL/Services/CsvParserService/CsvParserService.cs]: nit: I would add new line at least before the return for better readability

File: FundParser/BL/Services/DownloaderService/DownloaderService.cs [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+                var response = await _client.SendAsync(request, cancellationToken);
+                if (!response.IsSuccessStatusCode)
+                {
+                    await _logger.LogError($"Failed to download csv, server responded with status code: {response.StatusCode}",
+                        nameof(DownloaderService), cancellationToken);
+                    return null;
+                }
+                using Stream responseStream = await response.Content.ReadAsStreamAsync(cancellationToken);
+                using StreamReader reader = new(responseStream);
+                return await reader.ReadToEndAsync(cancellationToken);
```
Comment (adlermichal) [inline FundParser/BL/Services/DownloaderService/DownloaderService.cs:38]: nit: I would add a new line at least before the return for better readability

File: FundParser/BL/Services/FundCsvService/FundCsvService.cs [hunk 9]
```diff
@@ -1,47 +1,57 @@
...
     }
 
     public async Task<int> UpdateHoldings(CancellationToken cancellationToken = default)
     {
-        // TODO: move this to config
-        var url = "http://ark-funds.com/wp-content/uploads/funds-etf-csv/ARK_INNOVATION_ETF_ARKK_HOLDINGS.csv";
-        var rows = await _csvDownloader.DownloadAndParse(url, cancellationToken);
-
-        if (rows == null)
+        var url = _configuration.GetRequiredSection("CsvParserUrl").Value;
```
Comment (adlermichal) [inline FundParser/BL/Services/FundCsvService/FundCsvService.cs]: `CsvParserUrl` should be a constant, and it's also used in the error handling

File: FundParser/BL/Services/FundCsvService/FundCsvService.cs [hunk 10]
```diff
@@ -1,47 +1,57 @@
...
+        if (url is null)
         {
-            await _logger.LogError("Unable to download csv from the API", nameof(FundCsvService));
-            throw new Exception("Failed to download csv");
+            await _logger.LogError("CsvParserUrl is not set in the configuration", nameof(FundCsvService), cancellationToken);
+            throw new Exception("CsvParserUrl is not set in the configuration");
         }
```
Comment (adlermichal) [inline FundParser/BL/Services/FundCsvService/FundCsvService.cs:54]: nit: Add new line

File: FundParser/BL/Services/FundCsvService/FundCsvService.cs [hunk 11]
```diff
@@ -50,7 +60,7 @@ public async Task<int> UpdateHoldings(CancellationToken cancellationToken = defa
             }
             catch (Exception e)
             {
-                await _logger.LogError($"Unable to proccess csv row from the API {row}, thrown exception {e.Message}", nameof(FundCsvService));
+                await _logger.LogError($"Unable to proccess csv row from the API {row}, thrown exception {e.Message}", nameof(FundCsvService), cancellationToken);
             }
         }
```
Comment (adlermichal) [inline FundParser/BL/Services/FundCsvService/FundCsvService.cs:75]: nit: I would add a new line before the return for better readability

File: FundParser/BL/Services/FundCsvService/FundCsvService.cs [hunk 12]
```diff
@@ -1,47 +1,57 @@
...
-        var rows = await _csvDownloader.DownloadAndParse(url, cancellationToken);
-
-        if (rows == null)
+        var url = _configuration.GetRequiredSection("CsvParserUrl").Value;
+        if (url is null)
         {
-            await _logger.LogError("Unable to download csv from the API", nameof(FundCsvService));
-            throw new Exception("Failed to download csv");
+            await _logger.LogError("CsvParserUrl is not set in the configuration", nameof(FundCsvService), cancellationToken);
+            throw new Exception("CsvParserUrl is not set in the configuration");
```
Comment (adlermichal) [inline FundParser/BL/Services/FundCsvService/FundCsvService.cs]: Do we want to handle these exceptions in the controller (eg gracefully return 500) or do we just let them go through?

File: FundParser/BL/Services/FundCsvService/FundCsvService.cs [hunk 13]
```diff
@@ -1,47 +1,57 @@
...
         }
-
+        var csvString = await _downloaderService.DownloadTextFileAsStringAsync(url, _csvRequestHeaders, cancellationToken)
+            ?? throw new Exception("Failed to download csv");
+        var csvRows = _csvParser.ParseString(csvString, cancellationToken) ?? throw new Exception("Failed to parse csv");
```
Comment (adlermichal) [inline FundParser/BL/Services/FundCsvService/FundCsvService.cs]: nit: add new line

File: FundParser/BL/Services/FundCsvService/FundCsvService.cs [hunk 14]
```diff
@@ -50,7 +60,7 @@ public async Task<int> UpdateHoldings(CancellationToken cancellationToken = defa
             }
             catch (Exception e)
             {
-                await _logger.LogError($"Unable to proccess csv row from the API {row}, thrown exception {e.Message}", nameof(FundCsvService));
+                await _logger.LogError($"Unable to proccess csv row from the API {row}, thrown exception {e.Message}", nameof(FundCsvService), cancellationToken);
```
Comment (adlermichal) [inline FundParser/BL/Services/FundCsvService/FundCsvService.cs]: Does the `{row}` actually do anything without overridden `ToString`?

File: FundParser/BL/Services/FundCsvService/FundCsvService.cs [hunk 15]
```diff
@@ -59,9 +69,7 @@ public async Task<int> UpdateHoldings(CancellationToken cancellationToken = defa
     private async Task ProcessRow(FundCsvRow row, CancellationToken cancellationToken)
     {
         var holding = ParseFund(row);
-
```
Comment (adlermichal) [inline FundParser/BL/Services/FundCsvService/FundCsvService.cs:62]: nit: I would keep this space to visually separate the parsing and db handling

File: FundParser/BL/Services/FundCsvService/FundCsvService.cs [hunk 16]
```diff
@@ -88,7 +94,6 @@ private static AddHoldingDTO ParseFund(FundCsvRow row)
             Weight = decimal.Parse(row.weight.TrimEnd('%')),
             Date = DateTime.ParseExact(row.date, "MM/dd/yyyy", CultureInfo.InvariantCulture),
         };
-
         return holding;
```
Comment (adlermichal) [inline FundParser/BL/Services/FundCsvService/FundCsvService.cs:112]: nit: Imo all these new lines between model creation and return improved readability

File: FundParser/BL/Services/HoldingService/HoldingService.cs [hunk 17]
```diff
@@ -45,7 +45,7 @@ public async Task<HoldingDTO> AddHolding(AddHoldingDTO holding, CancellationToke
                 Date = holding.Date,
                 Weight = holding.Weight,
                 Shares = holding.Shares,
-            });
+            }, cancellationToken);
```
Comment (adlermichal) [inline FundParser/BL/Services/HoldingService/HoldingService.cs:48]: Good catch

File: FundParser/BL/Services/LoggingService/LoggingService.cs [hunk 18]
```diff
@@ -1,20 +1,19 @@
...
-        private readonly FundParserDbContext _context;
-
-        public LoggingService(IRepository<Log> logRepository,
-             FundParserDbContext isolatedContext)
+        private readonly string _logFolderPath;
+        public LoggingService(IConfiguration configuration)
         {
-            _logRepository = logRepository;
-            _context = isolatedContext;
+            _logFolderPath = configuration.GetRequiredSection("LogFolderPath").Value
```
Comment (adlermichal) [inline FundParser/BL/Services/LoggingService/LoggingService.cs]: As before, I would make `LogFolderPath` constant

File: FundParser/BL/Services/LoggingService/LoggingService.cs [hunk 19]
```diff
@@ -34,8 +33,8 @@ public async Task LogError(string message, string source, CancellationToken canc
 
         public async Task Log(string message, string source, LogLevel severity = LogLevel.None, CancellationToken cancellationToken = default)
```
Comment (adlermichal) [inline FundParser/BL/Services/LoggingService/LoggingService.cs]: The async/await combo can be omitted in all these methods if you return the `Log`, eg:
```
        public Task LogInformation(string message, string source, CancellationToken cancellationToken = default)
        {
            return Log(message, source, LogLevel.Information, cancellationToken);
        }
```
Not sure if is more readable though (or better)

File: FundParser/BL/Services/LoggingService/LoggingService.cs [hunk 20]
```diff
@@ -34,8 +33,8 @@ public async Task LogError(string message, string source, CancellationToken canc
...
         {
-            await _logRepository.Insert(new Log { Message = message, Source = source, Severity = severity }, cancellationToken);
-            await _context.SaveChangesAsync(cancellationToken);
+            var logFilePath = Path.Combine(_logFolderPath, $"{DateTime.Now:yyyy-MM-dd}.log");
+            await File.AppendAllTextAsync(logFilePath, $"{DateTime.Now:HH:mm:ss} {severity}: {source}: {message}\n", cancellationToken);
```
Comment (adlermichal) [inline FundParser/BL/Services/LoggingService/LoggingService.cs]: Use `Environment.NewLine` instead of `\n`

File: FundParser/BL/Services/DownloaderService/DownloaderService.cs [hunk 21]
```diff
@@ -0,0 +1,52 @@
+﻿using FundParser.BL.Services.LoggingService;
+
+namespace FundParser.BL.Services.DownloaderService
+{
+    public class DownloaderService : IDownloaderService
+    {
+        private readonly HttpClient _client;
+        private readonly ILoggingService _logger;
+
+        public DownloaderService(ILoggingService logger) : this(logger, new HttpClientHandler())
```
Comment (adlermichal) [inline FundParser/BL/Services/DownloaderService/DownloaderService.cs]: Since this is registered as singleton, we should ensure we don't run into issues with the HttpClient, see:
https://learn.microsoft.com/en-us/dotnet/fundamentals/networking/http/httpclient-guidelines#recommended-use
https://stackoverflow.com/questions/70924783/httpclient-inject-into-singleton

File: FundParser/DAL/Models/Company.cs [hunk 22]
```diff
@@ -4,15 +4,15 @@ namespace FundParser.DAL.Models
...
         [Required]
-        public string Name { get; set; }
+        public string Name { get; set; } = null!;
 
         [Required]
-        public string Cusip { get; set; }
+        public string Cusip { get; set; } = null!;
 
-        public string Ticker { get; set; }
+        public string Ticker { get; set; } = null!;
```
Comment (adlermichal) [inline FundParser/DAL/Models/Company.cs]: I think ticker is possibly nullable, see previous comment in `AddCompanyDTO`

File: FundParser/DAL/Models/FundCsvRow.cs [hunk 23]
```diff
@@ -5,26 +5,26 @@ namespace FundParser.DAL.Models;
 public class FundCsvRow
 {
     [Name("date")]
-    public string date { get; set; }
+    public string date { get; set; } = null!;
```
Comment (adlermichal) [inline FundParser/DAL/Models/FundCsvRow.cs]: Since we went with attributes for proper names here, can you also rename the properties to match conventions and start with uppercase? (or I can do it in reformat PR later on, just let me know so we don't forget)

File: FundParser/DAL/UnitOfWork/UnitOfWork.cs [hunk 24]
```diff
@@ -32,6 +32,7 @@ public class UnitOfWork : IUnitOfWork
     public void Dispose()
     {
         _context.Dispose();
+        GC.SuppressFinalize(this);
```
Comment (adlermichal) [inline FundParser/DAL/UnitOfWork/UnitOfWork.cs:35]: Why was this added?

File: FundParser/FundParser.App.Tests/FundParser.App.Tests.csproj
Comment (adlermichal) [inline FundParser/FundParser.App.Tests/FundParser.App.Tests.csproj:1]: Do we need this project at the moment? I think we can add it back when it's actually needed

File: FundParser/FundParser.BL.Tests/CsvParserTests.cs [hunk 25]
```diff
@@ -0,0 +1,73 @@
...
+        [SetUp]
+        public void Setup()
+        {
+            logger = new Mock<ILoggingService>();
+            parser = new CsvParserService<BasicCsvFormat>(logger.Object);
+            czCulture = new CultureInfo("cs-CZ");
+        }
+
+        [Test]
+        public void CsvParser_ParseString_BasicCsvCorrect()
```
Comment (adlermichal) [inline FundParser/FundParser.BL.Tests/CsvParserTests.cs]: I would go with naming: `ParseString_<Simple/Basic>Csv_ReturnsCorrectResult`

File: FundParser/FundParser.BL.Tests/CsvParserTests.cs [hunk 26]
```diff
@@ -0,0 +1,73 @@
...
+            {
+                Assert.That(result[0].DateTime, Is.EqualTo(DateTime.Parse("18/04/2023", czCulture)));
+                Assert.That(result[0].String, Is.EqualTo("ARKK2"));
+                Assert.That(result[0].Decimal, Is.EqualTo(1.21));
+                Assert.That(result[0].Int, Is.EqualTo(10000));
+            });
+        }
+
+        [Test]
+        public void CsvParser_ParseString_BasicCsvInvalidHeader()
```
Comment (adlermichal) [inline FundParser/FundParser.BL.Tests/CsvParserTests.cs]: Here it could be something like `ParseString_InvalidHeader_ReturnsNull`

File: FundParser/FundParser.BL.Tests/CsvParserTests.cs [hunk 27]
```diff
@@ -0,0 +1,73 @@
...
+        {
+            var basicCsv = "DateTime,String,Int\n" +
+                      "04/04/2024,ARKK,1.23\n" +
+                      "04/18/2023,ARKK2,1.21,10000\n" +
+                      "04/18/2024,ARKK3,1.25,420,Extra\n";
+            var result = parser.ParseString(basicCsv)?.ToList();
+            Assert.That(result, Is.EqualTo(null));
```
Comment (adlermichal) [inline FundParser/FundParser.BL.Tests/CsvParserTests.cs]: We should/could check that the logger received a call to `LogError`.

File: FundParser/FundParser.BL.Tests/CsvParserTests.cs [hunk 28]
```diff
@@ -0,0 +1,73 @@
...
+        {
+            logger = new Mock<ILoggingService>();
+            parser = new CsvParserService<BasicCsvFormat>(logger.Object);
+            czCulture = new CultureInfo("cs-CZ");
+        }
+
+        [Test]
+        public void CsvParser_ParseString_BasicCsvCorrect()
+        {
+            var basicCsv = "DateTime,String,Decimal,Int\n" +
```
Comment (adlermichal) [inline FundParser/FundParser.BL.Tests/CsvParserTests.cs]: Maybe create a private method like `GetCsvString` that will take `IEnumerable<String> rows` and then do `string.Join(Environment.NewLine, rows)`? Can be reused for all these tests

File: FundParser/FundParser.BL.Tests/CsvParserTests.cs [hunk 29]
```diff
@@ -0,0 +1,73 @@
...
+using Moq;
+using System.Globalization;
+
+namespace FundParser.BL.Tests
+{
+    public class CsvParserTests
+    {
+        private CsvParserService<BasicCsvFormat> parser;
+        CultureInfo czCulture;
+        private Mock<ILoggingService> logger;
```
Comment (adlermichal) [inline FundParser/FundParser.BL.Tests/CsvParserTests.cs]: We are naming private fields with `_` prefix per agreement

File: FundParser/FundParser.BL.Tests/CsvParserTests.cs [hunk 30]
```diff
@@ -0,0 +1,73 @@
...
+            czCulture = new CultureInfo("cs-CZ");
+        }
+
+        [Test]
+        public void CsvParser_ParseString_BasicCsvCorrect()
+        {
+            var basicCsv = "DateTime,String,Decimal,Int\n" +
+                      "04/04/2024,ARKK,1.23,69\n" +
+                      "04/18/2024,ARKK2,1.25,420\n";
+            var result = parser.ParseString(basicCsv)?.ToList();
```
Comment (adlermichal) [inline FundParser/FundParser.BL.Tests/CsvParserTests.cs]: Add new lines between the arrange, act, and assert phases for better readability.

File: FundParser/FundParser.BL.Tests/DownloaderTests.cs
Comment (adlermichal) [inline FundParser/FundParser.BL.Tests/DownloaderTests.cs:1]: Wrong file name, should be `DownloaderServiceTests`

File: FundParser/FundParser.BL.Tests/DownloaderTests.cs [hunk 31]
```diff
@@ -0,0 +1,138 @@
+using FundParser.BL.Services.DownloaderService;
+using FundParser.BL.Services.LoggingService;
+using FundParser.BL.Tests.Helpers;
+using Moq;
+using NUnit.Framework.Internal;
+
+namespace FundParser.BL.Tests
+{
+    public class DownloaderTests
```
Comment (adlermichal) [inline FundParser/FundParser.BL.Tests/DownloaderTests.cs]: Also fix class name

File: FundParser/FundParser.BL.Tests/DownloaderTests.cs [hunk 32]
```diff
@@ -0,0 +1,138 @@
...
+    {
+        private const string urlValid = "https://www.seznam.cz/";
```
Comment (adlermichal) [inline FundParser/FundParser.BL.Tests/DownloaderTests.cs]: Constants should be pascal case

File: FundParser/FundParser.BL.Tests/DownloaderTests.cs [hunk 33]
```diff
@@ -0,0 +1,138 @@
...
+        private const string urlInvalid = "invalid url";
+        private const string testString = "test, test2, test3";
+        private Mock<ILoggingService> logger;
```
Comment (adlermichal) [inline FundParser/FundParser.BL.Tests/DownloaderTests.cs]: `_` prefix for private fields

File: FundParser/FundParser.BL.Tests/DownloaderTests.cs [hunk 34]
```diff
@@ -0,0 +1,138 @@
...
+using Moq;
+using NUnit.Framework.Internal;
+
+namespace FundParser.BL.Tests
+{
+    public class DownloaderTests
+    {
+        private const string urlValid = "https://www.seznam.cz/";
+        private const string urlInvalid = "invalid url";
+        private const string testString = "test, test2, test3";
```
Comment (adlermichal) [inline FundParser/FundParser.BL.Tests/DownloaderTests.cs]: If these are used only in one test, move them to respective test methods.

File: FundParser/FundParser.BL.Tests/DownloaderTests.cs [hunk 35]
```diff
@@ -0,0 +1,138 @@
...
+                    Content = new StringContent(testString)
+                });
+
+            // Act
+            var downloaderService = new DownloaderService(logger.Object, new MockHttpMessageHandler(mockMessageHandler.Object));
+            var csvString = await downloaderService.DownloadTextFileAsStringAsync(urlValid, validHeaders, new CancellationToken());
+
+            // Assert
+            Assert.That(csvString, Is.EqualTo(testString));
+            try
```
Comment (adlermichal) [inline FundParser/FundParser.BL.Tests/DownloaderTests.cs]: Why catch? We want the test to fail anyway, right? I think it is completely valid to let it end with MockException rather than bloating the code with try-catches everytime we want to call `Verify`

File: FundParser/FundParser.BL.Tests/CsvParserTests.cs [hunk 36]
```diff
@@ -0,0 +1,73 @@
...
+
+namespace FundParser.BL.Tests
+{
+    public class CsvParserTests
+    {
+        private CsvParserService<BasicCsvFormat> parser;
+        CultureInfo czCulture;
+        private Mock<ILoggingService> logger;
+
+        [SetUp]
```
Comment (adlermichal) [inline FundParser/FundParser.BL.Tests/CsvParserTests.cs]: Btw we can introduce class for each tested method, and annotate those with `TestFixture` attribute

File: FundParser/FundParser.BL.Tests/CsvParserTests.cs [hunk 37]
```diff
@@ -0,0 +1,73 @@
+﻿using FundParser.BL.Services.CsvParserService;
+using FundParser.BL.Services.LoggingService;
+using FundParser.BL.Tests.Helpers.CsvFormats;
+using Moq;
+using System.Globalization;
+
+namespace FundParser.BL.Tests
+{
+    public class CsvParserTests
```
Comment (adlermichal) [inline FundParser/FundParser.BL.Tests/CsvParserTests.cs]: Add `TestFixture` attribute

File: FundParser/FundParser.BL.Tests/DownloaderTests.cs [hunk 38]
```diff
@@ -0,0 +1,138 @@
...
+                ), It.IsAny<CancellationToken>()), Times.Once);
+            }
+            catch (MockException e)
+            {
+                Assert.Fail(e.Message);
+            }
+        }
+
+        [Test]
+        public async Task DownloaderService_DownloadTextFile_InvalidLink()
```
Comment (adlermichal) [inline FundParser/FundParser.BL.Tests/DownloaderTests.cs]: I personally prefer invalid test cases first, and then valid ones. But we can discuss it and agree on valid, then invalid, of course.

File: FundParser/FundParser.BL.Tests/DownloaderTests.cs [hunk 39]
```diff
@@ -0,0 +1,138 @@
...
+
+        [SetUp]
+        public void Setup()
+        {
+            logger = new Mock<ILoggingService>();
+            mockMessageHandler = new Mock<IMockHttpMessageHandler>();
+        }
+
+        [Test]
+        public async Task DownloaderService_DownloadTextFile_Success()
```
Comment (adlermichal) [inline FundParser/FundParser.BL.Tests/DownloaderTests.cs]: Test names should contain full method name, also remove the class name from method name?

File: FundParser/FundParser.BL.Tests/CsvParserTests.cs (hunk 25 above)
Comment (adlermichal) [reply FundParser/FundParser.BL.Tests/CsvParserTests.cs]: Following the `Method_Scenario/What_ExpectedResult` pattern, see https://learn.microsoft.com/en-us/dotnet/core/testing/unit-testing-best-practices#naming-your-tests

File: FundParser/FundParser.BL.Tests/CsvParserTests.cs (hunk 29 above)
Comment (adlermichal) [inline FundParser/FundParser.BL.Tests/CsvParserTests.cs]: I would maybe name the mocks like so:  `_loggerMock` so that it is clear it is a mock and not the actual implementation.

File: FundParser/FundParser.BL.Tests/CsvParserTests.cs (hunk 37 above)
Comment (adlermichal) [reply FundParser/FundParser.BL.Tests/CsvParserTests.cs]: And should be named (along with file) `CsvParserServiceTests`

File: FundParser/FundParser.BL.Tests/DownloaderTests.cs [hunk 40]
```diff
@@ -0,0 +1,138 @@
...
+        {
+            // Setup
+            HttpRequestMessage request = new(HttpMethod.Get, urlValid);
+            foreach (var header in validHeaders)
+            {
+                request.Headers.Add(header.Item1, header.Item2);
+            }
+            mockMessageHandler.Setup(Setup => Setup.SendAsync(It.IsAny<HttpRequestMessage>(), It.IsAny<CancellationToken>()))
+                .ReturnsAsync(new HttpResponseMessage(System.Net.HttpStatusCode.OK)
```
Comment (adlermichal) [inline FundParser/FundParser.BL.Tests/DownloaderTests.cs]: Add using for `System.Net`

File: FundParser/FundParser.BL.Tests/DownloaderTests.cs [hunk 41]
```diff
@@ -0,0 +1,138 @@
...
+                logger.Verify(logger => logger.LogError(It.IsAny<string>(), It.IsAny<string>(), It.IsAny<CancellationToken>()), Times.Once);
+            } 
+            catch (MockException e)
+            {
+                Assert.Fail(e.Message);
+            }
+        }
+
+        [Test]
+        public async Task DownloaderService_DownloadTextFile_BadResponse()
```
Comment (adlermichal) [inline FundParser/FundParser.BL.Tests/DownloaderTests.cs]: Should we also have a test here for a case when HttpClient.SendAsync throws?

File: FundParser/FundParser.BL.Tests/Helpers/CsvFormats/BasicCsvFormat.cs [hunk 42]
```diff
@@ -0,0 +1,16 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace FundParser.BL.Tests.Helpers.CsvFormats
+{
+    public class BasicCsvFormat
+    {
+        public DateTime DateTime { get; set; }
```
Comment (adlermichal) [inline FundParser/FundParser.BL.Tests/Helpers/CsvFormats/BasicCsvFormat.cs]: nit: add new lines between properties

File: FundParser/FundParser.BL.Tests/Helpers/MoqHttpMessageHandler.cs
Comment (adlermichal) [inline FundParser/FundParser.BL.Tests/Helpers/MoqHttpMessageHandler.cs:1]: Rename file to match class name

File: FundParser/FundParser.BL.Tests/Helpers/MoqHttpMessageHandler.cs [hunk 43]
```diff
@@ -0,0 +1,28 @@
...
+namespace FundParser.BL.Tests.Helpers
+{
+    public interface IMockHttpMessageHandler
+    {
+        Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken);
+    }
+
+    public class MockHttpMessageHandler : HttpMessageHandler
+    {
+        private readonly IMockHttpMessageHandler _realMockHandler;
```
Comment (adlermichal) [inline FundParser/FundParser.BL.Tests/Helpers/MoqHttpMessageHandler.cs]: Is  `_mockHandler` enough? Or is the `real` in the name important?

File: FundParser/BL/Services/DownloaderService/DownloaderService.cs (hunk 21 above)
Comment (adlermichal) [reply FundParser/BL/Services/DownloaderService/DownloaderService.cs]: Also do we need both constuctors to be public? Do we need both of them? Can one of them be internal? (probably the second one with `HttpMesageHandler` param, since I assume that is only for testing?)

**Review by adlermichal — CHANGES_REQUESTED** [review]
Good job @Qas1modo! I appreciate that you've written (probably the worst out of the entire project) tests for the downloader.

File: FundParser/FundParser.BL.Tests/Helpers/CsvFormats/BasicCsvFormat.cs (hunk 42 above)
Comment (adlermichal) [reply FundParser/FundParser.BL.Tests/Helpers/CsvFormats/BasicCsvFormat.cs]: And maybe rename the folder to Mocks rather than helpers?

File: FundParser/FundParser.BL.Tests/CsvParserTests.cs (hunk 37 above)
Comment (adlermichal) [reply FundParser/FundParser.BL.Tests/CsvParserTests.cs]: And maybe move this and downloader tests to `Services` folder?

File: FundParser/BL/Services/CsvParserService/CsvParserService.cs (hunk 4 above)
Comment (samuelgocnik) [reply FundParser/BL/Services/CsvParserService/CsvParserService.cs]: `CsvParsingService` sounds good imo

File: FundParser/BL/DTOs/AddCompanyDTO.cs (hunk 3 above)
Comment (Qas1modo) [reply FundParser/BL/DTOs/AddCompanyDTO.cs]: I have for now removed all null!. We can add required if we want to get rid of the warnings in future.

File: FundParser/BL/Services/FundCsvService/FundCsvService.cs (hunk 9 above)
Comment (Qas1modo) [reply FundParser/BL/Services/FundCsvService/FundCsvService.cs]: It is constant defined inside appsettings.json - We should get rid of the configuration inside the code. This is the example of it.

File: FundParser/BL/Services/FundCsvService/FundCsvService.cs (hunk 12 above)
Comment (Qas1modo) [reply FundParser/BL/Services/FundCsvService/FundCsvService.cs]: I think it is not a problem, if we let them through.

File: FundParser/BL/Services/FundCsvService/FundCsvService.cs (hunk 14 above)
Comment (Qas1modo) [reply FundParser/BL/Services/FundCsvService/FundCsvService.cs]: Added overriden ToString

File: FundParser/BL/Services/FundCsvService/FundCsvService.cs (hunk 16 above)
Comment (Qas1modo) [reply FundParser/BL/Services/FundCsvService/FundCsvService.cs:112]: I do not see any difference in readability. However, I added it back.

File: FundParser/BL/Services/LoggingService/LoggingService.cs (hunk 18 above)
Comment (Qas1modo) [reply FundParser/BL/Services/LoggingService/LoggingService.cs]: 4. Žádná konfigurace nebude hardcoded
I think this is what is meant by this requirement.

File: FundParser/BL/Services/DownloaderService/DownloaderService.cs (hunk 21 above)
Comment (Qas1modo) [reply FundParser/BL/Services/DownloaderService/DownloaderService.cs]: I have modifed the downloader, so it has only one constructor and adhere to best practises with HttpClients.

File: FundParser/DAL/Models/FundCsvRow.cs (hunk 23 above)
Comment (Qas1modo) [reply FundParser/DAL/Models/FundCsvRow.cs]: i have renamed it

File: FundParser/DAL/UnitOfWork/UnitOfWork.cs (hunk 24 above)
Comment (Qas1modo) [reply FundParser/DAL/UnitOfWork/UnitOfWork.cs:35]: https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1816

File: FundParser/FundParser.BL.Tests/DownloaderTests.cs (hunk 38 above)
Comment (Qas1modo) [reply FundParser/FundParser.BL.Tests/DownloaderTests.cs]: I will keep it this way. If you want you can change it latter

File: FundParser/BL/Services/FundCsvService/FundCsvService.cs (hunk 9 above)
Comment (adlermichal) [reply FundParser/BL/Services/FundCsvService/FundCsvService.cs]: I meant to extract the string into a constant and use it in the method call as well as in error messages. I will push the fix.

File: FundParser/BL/Services/LoggingService/LoggingService.cs (hunk 18 above)
Comment (adlermichal) [reply FundParser/BL/Services/LoggingService/LoggingService.cs]: It is. See the other comment. We didn't understand each other.

File: FundParser/BL/Services/DownloaderService/DownloaderService.cs (hunk 21 above)
Comment (adlermichal) [reply FundParser/BL/Services/DownloaderService/DownloaderService.cs]: Hmm, I don't know if it is correct. We don't set the lifetime anywhere, so I am not sure. I made the service register as transient just to be sure we don't capture the client for too long, but rather per request.

File: FundParser/DAL/Models/FundCsvRow.cs (hunk 23 above)
Comment (adlermichal) [reply FundParser/DAL/Models/FundCsvRow.cs]: thanks

File: FundParser/DAL/UnitOfWork/UnitOfWork.cs (hunk 24 above)
Comment (adlermichal) [reply FundParser/DAL/UnitOfWork/UnitOfWork.cs:35]: Oh, I see, thx. (VS also dislikes it, I just haven't noticed before and I genuinely didn't know about this.)

File: FundParser/FundParser.BL.Tests/DownloaderTests.cs (hunk 38 above)
Comment (adlermichal) [reply FundParser/FundParser.BL.Tests/DownloaderTests.cs]: I reordered it a bit (just because I placed invalid cases in repo tests first and it faster to change it here lol)

**Review by adlermichal — APPROVED** [review]
